File: packages/typescript-sdk/src/composition/wait.ts
import { TimeoutError } from "../errors";

const terminalRunStatuses = new Set([
  "completed",
  "failed",
  "timed_out",
  "crashed",
  "system_failed",
  "canceled",
  "expired",
  "dead_letter",
]);

const sleep = (ms: number): Promise<void> =>
  new Promise((resolve) => {
    setTimeout(resolve, ms);
  });

/**
 * Polling policy for {@link waitForRun}.
 */
export type WaitForRunOptions = {
  /** Maximum wait time before timing out. */
  readonly timeoutMs?: number;
  /** Initial polling delay in milliseconds. */
  readonly initialDelayMs?: number;
  /** Maximum polling delay in milliseconds. */
  readonly maxDelayMs?: number;
  /** Exponential backoff multiplier for polling delay. */
  readonly factor?: number;
  /** Optional custom terminal-status predicate. */
  readonly isTerminal?: (status: string | undefined) => boolean;
  /** Optional AbortSignal to cancel polling. */
  readonly signal?: AbortSignal;
};

type RunStatusResponse = {
  readonly status?: string;
};

type RunFetcher<TRun extends RunStatusResponse> = (input: {
  readonly pathParams: { readonly runID: string };
}) => Promise<TRun>;

/**
 * Polls `getRun` until the run reaches a terminal status or times out.
 */
export const waitForRun = async <TRun extends RunStatusResponse>(
  getRun: RunFetcher<TRun>,
  runID: string,
  options?: WaitForRunOptions
): Promise<TRun> => {
  const timeoutMs = options?.timeoutMs ?? 60_000;
  const factor = options?.factor ?? 1.5;
  const maxDelayMs = options?.maxDelayMs ?? 10_000;

  let delayMs = options?.initialDelayMs ?? 500;
  const startedAt = Date.now();

  for (;;) {
    if (options?.signal?.aborted) {
      throw options.signal.reason ?? new Error("waitForRun aborted");
    }

    const run = await getRun({ pathParams: { runID } });
    const status = run.status;

    const isTerminal =
      options?.isTerminal?.(status) ??
      (status !== undefined && terminalRunStatuses.has(status));

    if (isTerminal) {
      return run;
    }

    if (Date.now() - startedAt > timeoutMs) {
      throw new TimeoutError({
        message: `waitForRun timed out after ${timeoutMs}ms for run ${runID}`,
        runId: runID,
        elapsedMs: Date.now() - startedAt,
      });
    }

    await sleep(delayMs);
    delayMs = Math.min(maxDelayMs, Math.max(1, Math.round(delayMs * factor)));
  }
};

File: packages/ruby-sdk/lib/strait/composition/wait.rb
# frozen_string_literal: true

module Strait
  module Composition
    WaitForRunOptions = Struct.new(
      :timeout_ms, :initial_delay_ms, :max_delay_ms, :factor, :is_terminal,
      keyword_init: true
    ) do
      def initialize(**kwargs)
        super
        self.timeout_ms ||= 300_000
        self.initial_delay_ms ||= 500
        self.max_delay_ms ||= 10_000
        self.factor ||= 1.5
        self.is_terminal ||= ->(status) { Strait::FSM.terminal_run_status?(status) }
      end
    end

    def self.wait_for_run(run_id, get_run:, get_status:, opts: nil)
      opts ||= WaitForRunOptions.new
      start = Process.clock_gettime(Process::CLOCK_MONOTONIC, :millisecond)
      delay = opts.initial_delay_ms

      loop do
        run = get_run.call(run_id)
        status = get_status.call(run)
        return run if opts.is_terminal.call(status)

        elapsed = Process.clock_gettime(Process::CLOCK_MONOTONIC, :millisecond) - start
        if elapsed >= opts.timeout_ms
          raise Strait::TimeoutError.new(
            "timed out waiting for run #{run_id}",
            run_id: run_id,
            elapsed_ms: elapsed
          )
        end

        sleep(delay / 1000.0)
        delay = [delay * opts.factor, opts.max_delay_ms].min
      end
    end
  end
end
